File: ts/packages/domain_ai/src/nodes/nodeLoadResults.ts
import { Runnable, RunnableConfig } from '@langchain/core/runnables';
import { Annotation } from "@langchain/langgraph";
import { Storage } from '@google-cloud/storage';
import { AIMessage } from '@langchain/core/messages';
import { registerNode, BaseStateSpec } from "./nodeUtils.js";

const storage = new Storage({
    credentials: {
        client_email: process.env.GCP_CLIENT_EMAIL,
        private_key: process.env.GCP_PRIVATE_KEY?.replace(/\\n/g, '\n'),
        project_id: process.env.GCP_PROJECT_ID,
    }
});
const bucketName = 'tp_resources';


export const NodeLoadResultsState = Annotation.Root({
    ligandDocking: Annotation<{ path: string, value: Map<string, any> }>({  // The key of the map should be a string holding a "row_identifier" and the value should be a custom data type that represents a PDBQT row.
        reducer: (prev, next) => next
    }),
    ligandPose: Annotation<{ path: string, value: Map<string, any> }>({  // Key and value of map to be determined.
        reducer: (prev, next) => next
    }),
});

type WithBaseState = typeof NodeLoadResultsState.State &
    ReturnType<typeof Annotation.Root<typeof BaseStateSpec>>["State"];


class _NodeLoadResults extends Runnable {

    static specs = {
        description: "Load inputs from the bucket",
        resources: {
            inputSpecs: ["ligand", "receptor", "box"],
            outputSpecs: [],
        },
        state: NodeLoadResultsState,
    }

    lc_namespace = []; // ATTENTION: Assigning an empty array for now to honor the contract with the Runnable class, which implements RunnableInterface.

    async invoke(state: WithBaseState, options?: Partial<RunnableConfig<Record<string, any>>>): Promise<Partial<WithBaseState>> {
        // ATTENTION_RONAK: Here we'll load the docking results from the bucket and into GraphState.

        try {
            if (!state.ligandDocking?.path || !state.ligandPose?.path) {
                throw new Error("Missing ligandDocking or ligandPose paths");
            }

            const resources = [
                { key: 'ligandDocking', path: state.ligandDocking.path },
                { key: 'ligandPose', path: state.ligandPose.path }
            ];

            const results: Record<string, any> = {};

            for (const { key, path } of resources) {
                try {
                    // Remove any bucket prefix if present
                    const blobName = path
                        .replace('tp_resources/', '');

                    console.log(`Attempting to download ${bucketName}/${blobName}`);

                    const [content] = await storage
                        .bucket(bucketName)
                        .file(blobName)
                        .download();

                    // Create value map with content
                    const valueMap = new Map<string, any>();
                    valueMap.set('path', path);
                    valueMap.set('content', content.toString());

                    results[key] = {
                        path,
                        value: valueMap
                    };

                    console.log(`Successfully loaded ${key}`);
                } catch (downloadError: any) {
                    console.error(`Download error for ${key}:`, downloadError);
                    // Preserve the original path even if download fails
                    results[key] = {
                        path,
                        value: new Map([['path', path], ['error', downloadError.message]])
                    };
                }
            }

            return {
                messages: [new AIMessage("Results loaded")],
                ligandDocking: results.ligandDocking,
                ligandPose: results.ligandPose
            };
        } catch (error: any) {
            console.error("Error in nodeLoadResults:", error);
            return {
                messages: [new AIMessage(`Error loading results: ${error.message}`)]
            };
        }
    }

}

export const NodeLoadResults = registerNode<typeof NodeLoadResultsState, typeof _NodeLoadResults>(_NodeLoadResults);

File: py/domain_ai/src/nodes/node_load_results.py
from typing import Dict, Any, Optional, Mapping, List
from langchain.schema import AIMessage
from pydantic import BaseModel, Field
from google.cloud import storage
import os
from .node_utils import register_node, BaseStateSpec
from langchain_core.runnables import Runnable, RunnableConfig
from google.oauth2 import service_account

# Initialize storage client
credentials_info = {
    "type": "service_account",
    "project_id": os.getenv("GCP_PROJECT_ID"),
    "private_key": os.getenv("GCP_PRIVATE_KEY").replace("\\n", "\n"),
    "client_email": os.getenv("GCP_CLIENT_EMAIL"),
    "auth_uri": "https://accounts.google.com/o/oauth2/auth",
    "token_uri": "https://oauth2.googleapis.com/token",
    "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
    "client_x509_cert_url": f"https://www.googleapis.com/robot/v1/metadata/x509/{os.getenv('GCP_CLIENT_EMAIL').replace('@', '%40')}",
    "universe_domain": "googleapis.com"
}
credentials = service_account.Credentials.from_service_account_info(credentials_info)
storage_client = storage.Client(credentials=credentials)
BUCKET_NAME = "tp_resources"

class DockingValue(BaseModel):
    path: str
    value: Dict[str, Any]

# Combined State Model
class NodeLoadResultsState(BaseModel):
    # Input fields
    ligand_docking: Optional[DockingValue] = Field(
        None, 
        description="PDBQT row data"
    )
    ligand_pose: Optional[DockingValue] = Field(
        None, 
        description="Pose data"
    )

    # Output fields
    ligand_docking_output: Optional[Dict[str, Dict[str, Any]]] = Field(
        None,
        description="The key should be a 'row_identifier' and the value should represent a PDBQT row"
    )
    ligand_pose_output: Optional[Dict[str, Dict[str, Any]]] = Field(
        None,
        description="Key and value of map to be determined"
    )

    # BaseStateSpec fields
    messages: Optional[List[AIMessage]] = None
    employment: Optional[Dict[str, Any]] = None

    class Config:
        arbitrary_types_allowed = True

class NodeLoadResults(Runnable):
    """Node to load docking results from the bucket."""

    meta = {
        "description": "Load docking results from the bucket.",
        "state_specs": {
            "inputs": ["ligand_docking", "ligand_pose"],
            "outputs": ["ligand_docking_output", "ligand_pose_output"],
        },
        "resource_specs": {
            "inputs": ["docking", "pose"],
            "outputs": [],
        },
    }

    def invoke(
        self,
        state: NodeLoadResultsState,
        config: Optional[RunnableConfig] = None,
        **kwargs: Any
    ) -> Dict[str, Any]:
        """Load the docking results from the bucket and into GraphState."""
        try:
            if not state.ligand_docking or not state.ligand_pose:
                raise ValueError("Missing ligand_docking or ligand_pose paths")

            resources = [
                {"key": "ligand_docking", "path": state.ligand_docking.path},
                {"key": "ligand_pose", "path": state.ligand_pose.path}
            ]

            results: Dict[str, Any] = {}

            for resource in resources:
                try:
                    # Remove any bucket prefix if present
                    blob_name = resource["path"].replace('tp_resources/', '')

                    print(f"Attempting to download {BUCKET_NAME}/{blob_name}")

                    # Download the content
                    bucket = storage_client.bucket(BUCKET_NAME)
                    blob = bucket.blob(blob_name)
                    content = blob.download_as_string().decode('utf-8')

                    # Create value map with content
                    value_map = {
                        "path": resource["path"],
                        "content": content
                    }

                    results[resource["key"]] = {
                        "path": resource["path"],
                        "value": value_map
                    }

                    print(f"Successfully loaded {resource['key']}")

                except Exception as download_error:
                    print(f"Download error for {resource['key']}:", download_error)
                    # Preserve the original path even if download fails
                    results[resource["key"]] = {
                        "path": resource["path"],
                        "value": {
                            "path": resource["path"],
                            "error": str(download_error)
                        }
                    }

            return {
                "messages": [AIMessage(content="Results loaded")],
                "ligand_docking_output": results["ligand_docking"],
                "ligand_pose_output": results["ligand_pose"]
            }

        except Exception as error:
            print("Error in node_load_results:", error)
            return {
                "messages": [AIMessage(content=f"Error loading results: {str(error)}")]
            }

# Register the node
node_load_results = register_node(NodeLoadResults)
